### DISCUSSION TITLE
Executing python scripts using Tauri

### DISCUSSION THREAD

**Discussion by jacobluke89** [tauri-apps/tauri#1645] [Q&A] [answered]
Hello there, 

I am new to Tauri, I am currently building an application using Electron, but finding it to be a power hungry beast and found this project, which seems to solve this issue! Anyway my main question; my application involves executing a number of python scripts as I need to analyse documents using NLP and python is the language for this. 

Does anyone know of a tutorial or package out there similar to this one https://efficientcoder.net/using-python-with-electron-tutorial/ for Electron that shows you how to execute a python script in Tauri. 

Thank you for your help and guidance. 

Jacob

**Comment by amrbashir** [comment, marked as answer]
you can add python runtime as a [sidecar](https://tauri.studio/en/docs/usage/guides/bundler/sidecar) to be bundled with your app and then use tauri's command api to spawn it, execute python code and get the result. 

If you don't want to bundle the python runtime you can just require your users to have python installed and added to the PATH in their OS.

**Reply by mihirsamdarshi** [reply]
Quick question regarding this, how would I go about creating a Flask backend for the application? I would assume that I would use something like pyInstaller to create a binary that could then be spawned by the Command API, and which could then run as a thread and accept requests at a locally defined endpoint?

Additionally in regards to the example Sidecar project in the Tauri repo, I was wondering if the [spawned command](https://github.com/tauri-apps/tauri/blob/268450b1329a4b55f2043890c565a8563f890c3a/examples/sidecar/src-tauri/src/main.rs#L23) was automatically exited when the main application exits? For example, if I have a backend Express server, would I need to handle quitting that process as well?

**Reply by amrbashir** [reply]
it will exit when your app exit, unless you kill it early

**Reply by Seliah** [reply]
Awesome feature, thanks!

**Reply by jmmaa** [reply]
I was able to use python binaries as I sidecar, but how do I exactly add python runtime to execute the scripts? especially when using virtualenvs?

**Reply by Seliah** [reply]
You can use [Pyinstaller](https://pyinstaller.readthedocs.io/en/stable/usage.html) in "onefile"-mode. It will generate an executable binary from your python script, libraries included. You can use that file as a sidecar.
Downside is that you have very long load times when starting the sidecar. I struggled a long time with this problem and gave up after a while. If someone has a tip for this I would appreciate it.

**Reply by shapiromatron** [reply]
The link above no longer works; sidecars are now described [here](https://tauri.app/v1/guides/building/sidecar) . Any chance someone is aware of a Tauri project or blogpost that describes this in greater detail?

**Reply by simonhyll** [reply]
I wouldn't recommend adding python to the bundle itself, it increases the size by quite a lot and tonnes of people already have Python installed anyway. Require your user to have it installed instead. Otherwise sidecaring it is what you'd have to do. If you want to use virtualenv I guess you can as a setup step in your app call the appropriate functions to create the venv somewhere on the users computer

**Reply by saml-dev** [reply]
@Seliah 's comment above is what I'm doing right now and it's working fine. I just run `pyinstaller -F path/to/my/file.py` and it gives me a binary. Then I configure that to be a sidecar. I'm starting the sidecar from TS in the client and it works well. I can also write to stdin and read from stdout to communicate with the python.

**Comment by ahmadyusufcodes** [comment]
Please someone should help create a boilerplate that works in 2024, I'm currently trying to set up a Tauri + NextJS + FastAPI, will come drop a link to template if I'm able to figure it out.
